File: csharp/1438.longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cs
/*
 * @lc app=leetcode id=1438 lang=csharp
 *
 * [1438] Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int LongestSubarray(int[] nums, int limit) {
        var minDeque = new LinkedList<int>();
        var maxDeque = new LinkedList<int>();
        int n = nums.Length, l = 0, ret = 0;
        for (int r = 0; r < n; ++r) {
            int curr = nums[r];
            while (minDeque.Count > 0 && nums[minDeque.Last.Value] >= curr) {
                minDeque.RemoveLast();
            }
            minDeque.AddLast(r);
            while (maxDeque.Count > 0 && nums[maxDeque.Last.Value] <= curr) {
                maxDeque.RemoveLast();
            }
            maxDeque.AddLast(r);
            while (nums[maxDeque.First.Value] - nums[minDeque.First.Value] > limit) {
                if (maxDeque.First.Value == l) {
                    maxDeque.RemoveFirst();
                }
                if (minDeque.First.Value == l) {
                    minDeque.RemoveFirst();
                }
                ++l;
            }
            ret = Math.Max(ret, r - l + 1);
        }
        return ret;
    }
}
// @lc code=end

File: java/1438.longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.java
/*
 * @lc app=leetcode id=1438 lang=java
 *
 * [1438] Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit
 */

import java.util.ArrayDeque;
// @lc code=start
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        ArrayDeque<Integer> max = new ArrayDeque<>(), min = new ArrayDeque<>();
        int n = nums.length, ret = 0, l = 0;
        for (int r = 0; r < n; ++r) {
            int num = nums[r];
            while (!max.isEmpty() && nums[max.getLast()] <= num) {
                max.removeLast();
            }
            max.addLast(r);
            while (!min.isEmpty() && nums[min.getLast()] >= num) {
                min.removeLast();
            }
            min.addLast(r);
            while (nums[max.getFirst()] - nums[min.getFirst()] > limit) {
                if (max.getFirst() == l) {
                    max.removeFirst();
                }
                if (min.getFirst() == l) {
                    min.removeFirst();
                }
                ++l;
            }
            ret = Math.max(ret, r - l + 1);
        }
        return ret;
    }
}
// @lc code=end

